File: py/day05.py
#!/usr/bin/env python3

from collections import Counter
import re
import sys

def line_points(x0, y0, x1, y1):
    dx = (x1 > x0) - (x1 < x0)
    dy = (y1 > y0) - (y1 < y0)
    while True:
        yield x0, y0
        if (x0, y0) == (x1, y1):
            break
        x0 += dx
        y0 += dy

def overlap_count(lines):
    counts = Counter(p for line in lines for p in line_points(*line))
    return sum(v > 1 for v in counts.values())

def main():
    lines = [tuple(map(int, re.findall("\d+", line))) for line in sys.stdin]
    straight = [(x0, y0, x1, y1) for x0, y0, x1, y1 in lines if x0 == x1 or y0 == y1]
    print(overlap_count(straight))
    print(overlap_count(lines))

if __name__ == "__main__":
    main()

File: src/bin/day05.rs
#![feature(io_read_to_string, iter_partition_in_place)]

use anyhow::{Context, Error, Result};
use std::{
    collections::HashMap,
    io::{read_to_string, stdin},
    str::FromStr,
};

fn main() -> Result<()> {
    let mut lines = parse_input(&read_to_string(&mut stdin())?)?;

    let straight = lines
        .iter_mut()
        .partition_in_place(|line| line.x0 == line.x1 || line.y0 == line.y1);
    let mut counts = HashMap::new();
    count_points(&mut counts, &lines[..straight]);
    println!("{}", counts.values().filter(|&&c| c > 1).count());
    count_points(&mut counts, &lines[straight..]);
    println!("{}", counts.values().filter(|&&c| c > 1).count());

    Ok(())
}

fn count_points(counts: &mut HashMap<(i32, i32), u32>, lines: &[Line]) {
    for line in lines {
        for p in line.points() {
            *counts.entry(p).or_default() += 1;
        }
    }
}

#[derive(Debug)]
struct Line {
    x0: i32,
    y0: i32,
    x1: i32,
    y1: i32,
}

impl Line {
    fn points(&self) -> impl Iterator<Item = (i32, i32)> + '_ {
        let dx = self.x1 - self.x0;
        let dy = self.y1 - self.y0;
        let len = std::cmp::max(dx.abs(), dy.abs());
        (0..=len).map(move |t| (self.x0 + sign(dx) * t, self.y0 + sign(dy) * t))
    }
}

fn sign(x: i32) -> i32 {
    (x > 0) as i32 - (x < 0) as i32
}

impl FromStr for Line {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let (p0, p1) = s.split_once(" -> ").context("invalid input line format")?;
        let parse_point = |p: &str| -> Result<(i32, i32)> {
            let (x, y) = p.split_once(",").context("invalid input line format")?;
            Ok((x.parse()?, y.parse()?))
        };
        let (x0, y0) = parse_point(p0)?;
        let (x1, y1) = parse_point(p1)?;
        Ok(Self { x0, y0, x1, y1 })
    }
}

fn parse_input(input: &str) -> Result<Vec<Line>> {
    input.lines().map(str::parse).collect()
}
